pool/heap.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pool
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/uber-go/tally"
+)
+
+// NewNativeHeap constructs a new BytesPool based on NativePool.
+func NewNativeHeap(b []Bucket, po ObjectPoolOptions) BytesPool {
+	if po == nil {
+		po = NewObjectPoolOptions()
+	}
+
+	var (
+		m        = po.MetricsScope()
+		ByteType = reflect.TypeOf((byte)(0))
+	)
+
+	sort.Sort(BucketByCapacity(b))
+
+	r := heap{m: heapMetrics{
+		overflows: m.Counter("overflows"),
+		misplaces: m.Counter("misplaces"),
+	}}
+
+	for _, cfg := range b {
+		m := m.SubScope(fmt.Sprintf("slot-%d", cfg.Capacity))
+
+		v := &slot{class: cfg.Capacity, opts: NativePoolOptions{
+			Size: uint(cfg.Count),
+			Type: reflect.ArrayOf(cfg.Capacity, ByteType),
+		}, m: slotMetrics{
+			used: m.Gauge("used"),
+			size: m.Gauge("size"),
+		}}
+
+		r.slots = append(r.slots, v)
+	}
+
+	return r
+}
+
+type heap struct {
+	slots []*slot
+
+	m heapMetrics
+}
+
+type heapMetrics struct {
+	overflows tally.Counter
+	misplaces tally.Counter
+}
+
+type slot struct {
+	sync.RWMutex
+
+	class int
+	opts  NativePoolOptions
+	pools []NativePool
+
+	m slotMetrics
+}
+
+type slotMetrics struct {
+	used tally.Gauge
+	size tally.Gauge
+}
+
+func (s *slot) get() interface{} {
+	if segment := s.getOr(s.RLocker(), func() interface{} {
```

Comment:
I guess if we later on add other heap types, like structs or whatnot, then it'd make sense to call it something else. I don't want to call it slice since it's ambiguous and might make one think that we're talking golang slices here, while we are not. Object is not really working too since it's an array of objects (in this case, bytes). So let's revisit it when adding other heap types.